diff -Naur a/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt b/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/CMakeLists.txt	2020-07-16 01:36:51.988582173 +0200
@@ -5,6 +5,7 @@
             DVDDemuxCDDA.cpp
             DVDDemuxClient.cpp
             DVDDemuxFFmpeg.cpp
+            DVDDemuxFFmpegArchive.cpp
             DVDDemuxUtils.cpp
             DVDDemuxVobsub.cpp
             DVDFactoryDemuxer.cpp)
@@ -16,6 +17,7 @@
             DVDDemuxCDDA.h
             DVDDemuxClient.h
             DVDDemuxFFmpeg.h
+            DVDDemuxFFmpegArchive.h
             DVDDemuxUtils.h
             DVDDemuxVobsub.h
             DVDFactoryDemuxer.h)
diff -Naur a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.cpp b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.cpp
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.cpp	1970-01-01 01:00:00.000000000 +0100
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.cpp	2020-07-16 01:36:51.988582173 +0200
@@ -0,0 +1,71 @@
+/*
+ *  Copyright (C) 2018 Arthur Liberman
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DVDDemuxFFmpegArchive.h"
+
+#include "cores/VideoPlayer/Interface/Addon/TimingConstants.h" // for DVD_NOPTS_VALUE
+#include "DVDDemuxUtils.h"
+#include "DVDInputStreams/DVDInputStream.h"
+#include "DVDInputStreams/DVDInputStreamFFmpegArchive.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+CDVDDemuxFFmpegArchive::CDVDDemuxFFmpegArchive() :
+    m_bIsOpening(false), m_seekOffset(0)
+{ }
+
+bool CDVDDemuxFFmpegArchive::Open(std::shared_ptr<CDVDInputStream> pInput, bool fileinfo)
+{
+  m_bIsOpening = true;
+  bool ret = CDVDDemuxFFmpeg::Open(pInput, fileinfo);
+  m_bIsOpening = false;
+  return ret;
+}
+
+bool CDVDDemuxFFmpegArchive::SeekTime(double time, bool backwards, double *startpts)
+{
+  if (!m_pInput || time < 0)
+    return false;
+
+  int whence = m_bIsOpening ? SEEK_CUR : SEEK_SET;
+  int64_t seekResult = m_pInput->Seek(static_cast<int64_t>(time), whence);
+  if (seekResult >= 0)
+  {
+    {
+      CSingleLock lock(m_critSection);
+      m_seekOffset = seekResult;
+    }
+
+    CLog::Log(LOGDEBUG, "Seek successful. m_seekOffset = %f, m_currentPts = %f, time = %f, backwards = %d, startptr = %f",
+      m_seekOffset, m_currentPts, time, backwards, startpts ? *startpts : 0);
+    return m_bIsOpening ? true : Reset();
+  }
+
+  CLog::Log(LOGDEBUG, "Seek failed. m_currentPts = %f, time = %f, backwards = %d, startptr = %f",
+    m_currentPts, time, backwards, startpts ? *startpts : 0);
+  return false;
+}
+
+DemuxPacket* CDVDDemuxFFmpegArchive::Read()
+{
+  DemuxPacket* pPacket = CDVDDemuxFFmpeg::Read();
+  if (pPacket)
+  {
+    CSingleLock lock(m_critSection);
+    pPacket->pts += m_seekOffset;
+    pPacket->dts += m_seekOffset;
+  }
+
+  return pPacket;
+}
+
+void CDVDDemuxFFmpegArchive::UpdateCurrentPTS()
+{
+  CDVDDemuxFFmpeg::UpdateCurrentPTS();
+  if (m_currentPts != DVD_NOPTS_VALUE)
+    m_currentPts += m_seekOffset;
+}
diff -Naur a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.h b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.h
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.h	1970-01-01 01:00:00.000000000 +0100
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpegArchive.h	2020-07-16 01:36:51.988582173 +0200
@@ -0,0 +1,24 @@
+/*
+ *  Copyright (C) 2018 Arthur Liberman
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "DVDDemuxFFmpeg.h"
+
+class CDVDDemuxFFmpegArchive : public CDVDDemuxFFmpeg
+{
+public:
+    CDVDDemuxFFmpegArchive();
+    virtual bool Open(std::shared_ptr<CDVDInputStream> pInput, bool fileinfo = false) override;
+    virtual bool SeekTime(double time, bool backwards = false, double* startpts = NULL) override;
+    virtual DemuxPacket* Read() override;
+protected:
+    virtual void UpdateCurrentPTS() override;
+
+    bool m_bIsOpening;
+    double m_seekOffset;
+};
diff -Naur a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.h	2020-07-16 01:39:23.712579646 +0200
@@ -83,7 +83,7 @@
   CDVDDemuxFFmpeg();
   ~CDVDDemuxFFmpeg() override;
 
-  bool Open(std::shared_ptr<CDVDInputStream> pInput, bool fileinfo);
+  virtual bool Open(std::shared_ptr<CDVDInputStream> pInput, bool fileinfo);
   void Dispose();
   bool Reset() override ;
   void Flush() override;
@@ -130,7 +130,7 @@
   void ResetVideoStreams();
   AVDictionary *GetFFMpegOptionsFromInput();
   double ConvertTimestamp(int64_t pts, int den, int num);
-  void UpdateCurrentPTS();
+  virtual void UpdateCurrentPTS();
   bool IsProgramChange();
   unsigned int HLSSelectProgram();
 
diff -Naur a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDFactoryDemuxer.cpp b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDFactoryDemuxer.cpp
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDFactoryDemuxer.cpp	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDFactoryDemuxer.cpp	2020-07-16 01:36:51.988582173 +0200
@@ -11,6 +11,7 @@
 #include "DVDInputStreams/DVDInputStream.h"
 
 #include "DVDDemuxFFmpeg.h"
+#include "DVDDemuxFFmpegArchive.h"
 #include "DVDDemuxBXA.h"
 #include "DVDDemuxCDDA.h"
 #include "DVDDemuxClient.h"
@@ -71,6 +72,17 @@
       return NULL;
   }
 
+  if (pInputStream->IsStreamType(DVDSTREAM_TYPE_PVR_ARCHIVE))
+  {
+    CLog::Log(LOGDEBUG, "DVDFactoryDemuxer: Create CDVDDemuxFFmpegArchive.");
+    std::unique_ptr<CDVDDemuxFFmpegArchive> demuxer(new CDVDDemuxFFmpegArchive());
+    if (demuxer->Open(pInputStream, fileinfo))
+      return demuxer.release();
+    else
+      return NULL;
+  }
+
+  CLog::Log(LOGDEBUG, "DVDFactoryDemuxer: Create CDVDDemuxFFmpeg.");
   std::unique_ptr<CDVDDemuxFFmpeg> demuxer(new CDVDDemuxFFmpeg());
   if (demuxer->Open(pInputStream, fileinfo))
     return demuxer.release();
diff -Naur a/xbmc/cores/VideoPlayer/DVDInputStreams/CMakeLists.txt b/xbmc/cores/VideoPlayer/DVDInputStreams/CMakeLists.txt
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/CMakeLists.txt	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/CMakeLists.txt	2020-07-16 01:36:51.988582173 +0200
@@ -1,6 +1,7 @@
 set(SOURCES DVDFactoryInputStream.cpp
             DVDInputStream.cpp
             DVDInputStreamFFmpeg.cpp
+            DVDInputStreamFFmpegArchive.cpp
             DVDInputStreamFile.cpp
             DVDInputStreamMemory.cpp
             DVDInputStreamNavigator.cpp
@@ -15,6 +16,7 @@
 set(HEADERS DVDFactoryInputStream.h
             DVDInputStream.h
             DVDInputStreamFFmpeg.h
+            DVDInputStreamFFmpegArchive.h
             DVDInputStreamFile.h
             DVDInputStreamMemory.h
             DVDInputStreamNavigator.h
diff -Naur a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp	2020-07-16 01:36:51.992582061 +0200
@@ -11,6 +11,7 @@
 #include "DVDInputStreamFile.h"
 #include "DVDInputStreamNavigator.h"
 #include "DVDInputStreamFFmpeg.h"
+#include "DVDInputStreamFFmpegArchive.h"
 #include "InputStreamAddon.h"
 #include "InputStreamMultiSource.h"
 #include "InputStreamPVRChannel.h"
@@ -29,6 +30,7 @@
 #include "ServiceBroker.h"
 #include "addons/binary-addons/BinaryAddonManager.h"
 #include "Util.h"
+#include "utils/log.h"
 
 
 std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVideoPlayer* pPlayer, const CFileItem &fileitem, bool scanforextaudio)
@@ -143,8 +145,18 @@
       }
     }
 
-    if (finalFileitem.IsType(".m3u8"))
-      return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
+    if (finalFileitem.IsType(".m3u8") || finalFileitem.IsType(".php"))
+    {
+      if (fileitem.IsPVRChannelWithArchive() || fileitem.IsEPGWithArchive())
+      {
+        CLog::Log(LOGDEBUG, "%s: CDVDInputStreamFFmpegArchive", __FUNCTION__);
+        return std::shared_ptr<CDVDInputStreamFFmpegArchive>(new CDVDInputStreamFFmpegArchive(finalFileitem));
+      }
+      else
+      {
+        return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
+      }
+    }
 
     if (finalFileitem.GetMimeType() == "application/vnd.apple.mpegurl")
       return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
@@ -154,6 +166,7 @@
   }
 
   // our file interface handles all these types of streams
+  CLog::Log(LOGDEBUG, "%s: All else failed, creating CDVDInputStreamFile", __FUNCTION__);
   return std::shared_ptr<CDVDInputStreamFile>(new CDVDInputStreamFile(finalFileitem,
                                                                       XFILE::READ_TRUNCATED |
                                                                       XFILE::READ_BITRATE |
diff -Naur a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.cpp b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.cpp
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.cpp	1970-01-01 01:00:00.000000000 +0100
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.cpp	2020-07-16 01:36:51.992582061 +0200
@@ -0,0 +1,69 @@
+/*
+ *  Copyright (C) 2018 Arthur Liberman
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "DVDInputStreamFFmpegArchive.h"
+#include "ServiceBroker.h"
+#include "addons/PVRClient.h"
+#include "pvr/PVRManager.h"
+#include "utils/log.h"
+
+CDVDInputStreamFFmpegArchive::CDVDInputStreamFFmpegArchive(const CFileItem& fileitem)
+  : CDVDInputStreamFFmpeg(fileitem),
+    m_client(CServiceBroker::GetPVRManager().GetClient(fileitem))
+{ }
+
+int64_t CDVDInputStreamFFmpegArchive::GetLength()
+{
+  int64_t ret = 0;
+  if (m_client && m_client->GetLiveStreamLength(ret) != PVR_ERROR_NO_ERROR)
+  {
+    Times times = {0};
+    if (GetTimes(times) && times.ptsEnd >= times.ptsBegin)
+      ret = static_cast<int64_t>(times.ptsEnd - times.ptsBegin);
+  }
+  return ret;
+}
+
+bool CDVDInputStreamFFmpegArchive::GetTimes(Times &times)
+{
+  bool ret = false;
+  PVR_STREAM_TIMES streamTimes = {0};
+  if (m_client && m_client->GetStreamTimes(&streamTimes) == PVR_ERROR_NO_ERROR)
+  {
+    times.startTime = streamTimes.startTime;
+    times.ptsStart = static_cast<double>(streamTimes.ptsStart);
+    times.ptsBegin = static_cast<double>(streamTimes.ptsBegin);
+    times.ptsEnd = static_cast<double>(streamTimes.ptsEnd);
+    ret = true;
+  }
+  return ret;
+}
+
+int64_t CDVDInputStreamFFmpegArchive::Seek(int64_t offset, int whence)
+{
+  int64_t iPosition = -1;
+  if (m_client)
+    m_client->SeekLiveStream(offset, whence, iPosition);
+  return iPosition;
+}
+
+CURL CDVDInputStreamFFmpegArchive::GetURL()
+{
+  if (m_client)
+  {
+    if (m_item.HasEPGInfoTag())
+      m_client->FillEpgTagStreamFileItem(m_item);
+    else if (m_item.HasPVRChannelInfoTag())
+      m_client->FillChannelStreamFileItem(m_item);
+  }
+  return CDVDInputStream::GetURL();
+}
+
+bool CDVDInputStreamFFmpegArchive::IsStreamType(DVDStreamType type) const
+{
+    return CDVDInputStream::IsStreamType(type) || type == DVDSTREAM_TYPE_PVR_ARCHIVE;
+}
diff -Naur a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.h b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.h
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.h	1970-01-01 01:00:00.000000000 +0100
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpegArchive.h	2020-07-16 01:36:51.992582061 +0200
@@ -0,0 +1,34 @@
+/*
+ *  Copyright (C) 2018 Arthur Liberman
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "DVDInputStreamFFmpeg.h"
+
+namespace PVR
+{
+  class CPVRClient;
+}
+
+class CDVDInputStreamFFmpegArchive
+  : public CDVDInputStreamFFmpeg
+  , public CDVDInputStream::ITimes
+{
+public:
+    explicit CDVDInputStreamFFmpegArchive(const CFileItem& fileitem);
+
+    virtual bool IsStreamType(DVDStreamType type) const override;
+
+    int64_t GetLength() override;
+    int64_t Seek(int64_t offset, int whence) override;
+    CURL GetURL() override;
+    CDVDInputStream::ITimes* GetITimes() override { return this; }
+    bool GetTimes(Times &times);
+
+protected:
+    std::shared_ptr<PVR::CPVRClient> m_client;
+};
diff -Naur a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h
--- a/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.h	2020-07-16 01:36:51.992582061 +0200
@@ -30,7 +30,8 @@
   DVDSTREAM_TYPE_BLURAY = 11,
   DVDSTREAM_TYPE_PVRMANAGER = 12,
   DVDSTREAM_TYPE_MULTIFILES = 13,
-  DVDSTREAM_TYPE_ADDON = 14
+  DVDSTREAM_TYPE_ADDON = 14,
+  DVDSTREAM_TYPE_PVR_ARCHIVE = 15
 };
 
 #define SEEK_POSSIBLE 0x10 // flag used to check if protocol allows seeks
@@ -169,7 +170,7 @@
    */
   virtual bool GetCacheStatus(XFILE::SCacheStatus *status) { return false; }
 
-  bool IsStreamType(DVDStreamType type) const { return m_streamType == type; }
+  virtual bool IsStreamType(DVDStreamType type) const { return m_streamType == type; }
   virtual bool IsEOF() = 0;
   virtual BitstreamStats GetBitstreamStats() const { return m_stats; }
 
diff -Naur a/xbmc/FileItem.cpp b/xbmc/FileItem.cpp
--- a/xbmc/FileItem.cpp	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/FileItem.cpp	2020-07-16 01:36:51.988582173 +0200
@@ -845,11 +845,27 @@
   return HasEPGInfoTag();
 }
 
+bool CFileItem::IsEPGWithArchive() const
+{
+  if (HasEPGInfoTag())
+  {
+    const std::shared_ptr<CPVRChannel> channel = CServiceBroker::GetPVRManager().ChannelGroups()->GetChannelForEpgTag(GetEPGInfoTag());
+    if (channel)
+      return channel->HasArchive();
+  }
+  return false;
+}
+
 bool CFileItem::IsPVRChannel() const
 {
   return HasPVRChannelInfoTag();
 }
 
+bool CFileItem::IsPVRChannelWithArchive() const
+{
+  return (HasPVRChannelInfoTag() && GetPVRChannelInfoTag()->HasArchive());
+}
+
 bool CFileItem::IsPVRRecording() const
 {
   return HasPVRRecordingInfoTag();
diff -Naur a/xbmc/FileItem.h b/xbmc/FileItem.h
--- a/xbmc/FileItem.h	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/FileItem.h	2020-07-16 01:36:51.988582173 +0200
@@ -223,7 +223,9 @@
   bool IsMusicDb() const;
   bool IsVideoDb() const;
   bool IsEPG() const;
+  bool IsEPGWithArchive() const;
   bool IsPVRChannel() const;
+  bool IsPVRChannelWithArchive() const;
   bool IsPVRRecording() const;
   bool IsUsablePVRRecording() const;
   bool IsDeletedPVRRecording() const;
diff -Naur a/xbmc/pvr/PVRManager.cpp b/xbmc/pvr/PVRManager.cpp
--- a/xbmc/pvr/PVRManager.cpp	2020-05-31 15:31:35.000000000 +0200
+++ b/xbmc/pvr/PVRManager.cpp	2020-07-16 01:36:51.992582061 +0200
@@ -960,12 +960,28 @@
   const CPVRClientPtr client = GetClient(fileItem);
   if (client)
   {
+    if (client->GetBackendName() == "IPTV Archive PVR Add-on")
+      client->CloseLiveStream();
     if (fileItem.IsPVRChannel())
       return client->FillChannelStreamFileItem(fileItem) == PVR_ERROR_NO_ERROR;
     else if (fileItem.IsPVRRecording())
       return client->FillRecordingStreamFileItem(fileItem) == PVR_ERROR_NO_ERROR;
     else if (fileItem.IsEPG())
-      return client->FillEpgTagStreamFileItem(fileItem) == PVR_ERROR_NO_ERROR;
+    {
+      PVR_ERROR error = client->FillEpgTagStreamFileItem(fileItem);
+      if (error == PVR_ERROR_NOT_IMPLEMENTED)
+      {
+        // PVR_ERROR_NOT_IMPLEMENTED is returned when Play EPG as Live TV is selected.
+        const std::shared_ptr<CPVRChannel> channel = CServiceBroker::GetPVRManager().ChannelGroups()->GetChannelForEpgTag(fileItem.GetEPGInfoTag());
+        if (channel)
+        {
+          fileItem = CFileItem(channel);
+          return client->FillChannelStreamFileItem(fileItem) == PVR_ERROR_NO_ERROR;
+        }
+      }
+
+      return error == PVR_ERROR_NO_ERROR;
+    }
   }
   return false;
 }
